src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,377 @@
+                    }
+                    return sb.ToString();
+                }
+            }
+            return base.ToString();
+        }
+
+        /// <summary>
+        /// Returns an enumerator over the <see cref="ReadOnlyBuffer{T}"/>
+        /// </summary>
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        public SequencePosition GetPosition(SequencePosition origin, long offset)
+        {
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+            return Seek(origin, BufferEnd, offset, false);
+        }
+
+        public bool TryGet(ref SequencePosition position, out ReadOnlyMemory<T> data, bool advance = true)
+        {
+            var result = TryGetBuffer(position, End, out data, out var next);
+            if (advance)
+            {
+                position = next;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// An enumerator over the <see cref="ReadOnlyBuffer{T}"/>
+        /// </summary>
+        public struct Enumerator
+        {
+            private readonly ReadOnlyBuffer<T> _readOnlyBuffer;
+            private SequencePosition _next;
+            private ReadOnlyMemory<T> _currentMemory;
+
+            /// <summary>
+            ///
+            /// </summary>
+            public Enumerator(ReadOnlyBuffer<T> readOnlyBuffer)
+            {
+                _readOnlyBuffer = readOnlyBuffer;
+                _currentMemory = default;
+                _next = readOnlyBuffer.Start;
+            }
+
+            /// <summary>
+            /// The current <see cref="Memory{Byte}"/>
+            /// </summary>
+            public ReadOnlyMemory<T> Current => _currentMemory;
+
+            /// <summary>
+            /// Moves to the next <see cref="Memory{Byte}"/> in the <see cref="ReadOnlyBuffer{T}"/>
+            /// </summary>
+            /// <returns></returns>
+            public bool MoveNext()
+            {
+                if (_next.Segment == null)
+                {
+                    return false;
+                }
+
+                return _readOnlyBuffer.TryGet(ref _next, out _currentMemory);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private ReadOnlyBuffer<T> SliceImpl(SequencePosition begin, SequencePosition end)
+        {
+            return new ReadOnlyBuffer<T>(
+                begin.Segment,
+                begin.Index & IndexBitMask | (Start.Index & ~IndexBitMask),
+                end.Segment,
+                end.Index & IndexBitMask | (End.Index & ~IndexBitMask)
+            );
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private BufferType GetBufferType()
+        {
+            // We take high order bits of two indexes index and move them
```

Comment:
typo

Reply:
are you saying the 2 should be a BufferTypeBitMask?

Reply:
Bitwise and with 2 is to keep only the second bit of `(uint)Start.Index >> 30`. We shift it by 30 to get MSB into bit 2 position, but bit 1 might be set so we need to clear it.

Reply:
Not really. I will think about it some more because I don't understand this method. If BufferTypeBitMask fits here, then it should be used instead of `2`

Reply:
> We take high order bits of two indexes index and move them

Extra word `index` needs to be removed